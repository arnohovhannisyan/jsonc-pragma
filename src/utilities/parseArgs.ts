import { IArguments } from "../models/arguments";

export function parseArgs(match: string | null): IArguments {
	if (!match) return {};

	const entries = [
		...new (typeof URLSearchParams === "undefined"
			? require("url").URLSearchParams
			: URLSearchParams)(match.replace(/\s(\w*?=)/, "&$1"))
	];

	return entries.reduce(
		(acc, [k, v]) => ({
			...acc,
			[k]: v
		}),
		{}
	);
}
