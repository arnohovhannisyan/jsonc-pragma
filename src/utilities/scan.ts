import matchBracket from "match-bracket";
import { ISection } from "../models/section";
import { parseArgs } from "./parseArgs";

function isRegular(line: string) {
	if (/[{[]$/.test(line.trim())) return false;

	return true;
}

/**
 * @param contents The contents of your JSON file
 */
export function scan(contents: string): ISection[] {
	const pragmaRegex = /\s*\/\/\s@(?<name>[a-zA-Z\d]+)(\s(?<args>.*))?$/;

	const lines = contents.split("\n");

	const linesWithPragmas = lines.filter(line => pragmaRegex.test(line));

	return linesWithPragmas.map<ISection>(line => {
		const startingLineNumber = lines.indexOf(line) + 1;
		const startingLine = lines[startingLineNumber];

		const groups = pragmaRegex.exec(line)!.groups!;

		if (isRegular(startingLine)) {
			return {
				start: startingLineNumber,
				end: startingLineNumber,
				args: parseArgs(groups.args),
				name: groups.name
			};
		}

		const startingBracketPos = {
			line: startingLineNumber + 1,
			cursor: startingLine.trimEnd().length
		};

		const endingLineNumber =
			matchBracket(contents.toString(), startingBracketPos).line - 1;

		return {
			start: startingLineNumber,
			end: endingLineNumber,
			args: parseArgs(groups.args),
			name: groups.name
		};
	});
}
