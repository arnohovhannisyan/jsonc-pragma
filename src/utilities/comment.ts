import { ISection } from "~/models";
import { scan } from "~/utilities/scan";

/**
 * @param contents The contents of your JSON file
 * @param selector A function that accepts a section and determines if it should be commented by returning true (comment) or false (don't comment).
 */
export function comment(
	contents: string | Buffer,
	selector: (section: ISection) => boolean = () => true
): string {
	const sections = scan(contents);

	const sectionsToComment = sections.filter(selector);

	const lines = contents.toString().split("\n");

	return lines
		.map((line, i) => {
			if (line.trimStart().startsWith("//")) return line;

			const section = sectionsToComment.find(s => i >= s.start && i <= s.end);

			if (section) {
				const firstLine = lines[section.start];
				const leadingSpaces = /^(\s*)/.exec(firstLine)?.[0].length ?? 0;

				return line.replace(new RegExp(`^(\\s{${leadingSpaces}})`), "$1// ");
			}

			return line;
		})
		.join("\n");
}
