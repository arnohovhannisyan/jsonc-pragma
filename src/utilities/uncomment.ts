import { ISection } from "../models/section";
import { scan } from "./scan";

/**
 * @param contents The contents of your JSON file
 * @param selector A function that accepts a section and determines if it should be uncommented by returning true (uncomment) or false (don't uncomment).
 */
export function uncomment(
	contents: string,
	selector: (section: ISection) => boolean = () => true
): string {
	const sections = scan(contents.toString().replace(/\/\/\s(?!@)/g, ""));

	const sectionsToUncomment = sections.filter(selector);

	const uncommented = contents
		.split("\n")
		.map((line, i) => {
			const shouldUncomment = sectionsToUncomment.some(
				section => i >= section.start && i <= section.end
			);

			if (shouldUncomment) return line.replace("// ", "");

			return line;
		})
		.join("\n");

	return uncommented;
}
