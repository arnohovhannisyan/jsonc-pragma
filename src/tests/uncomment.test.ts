import { uncomment } from "../utilities/uncomment";

test("basic functionality", () => {
	const input = `{
		// @foo bar=5
		// "key": true
	}`;

	expect(uncomment(input)).toMatchSnapshot();
});

test("multiple pragmas", () => {
	const input = `{
		// @foo bar=5
		// "key": true,

		// @baz bar=10
		// "key": true
	}`;

	expect(uncomment(input)).toMatchSnapshot();
});

test("selectors", () => {
	const input = `{
		// @foo bar=5
		// "key": true,

		// @baz bar=10
		// "key": true
	}`;

	expect(uncomment(input, s => Number(s.args.bar) > 5)).toMatchSnapshot();
});

test("objects", () => {
	const input = `{
		// @foo bar=5
		// "key": {
		// 	"key": true
		// }
	}`;

	expect(uncomment(input)).toMatchSnapshot();
});

test("arrays", () => {
	const input = `{
		// @foo bar=5
		// "key": [
		// 	{ "key": true }
		// ]
	}`;

	expect(uncomment(input)).toMatchSnapshot();
});

test("nested objects", () => {
	const input = `{
		// @foo bar=5
		// "key": {
		// 	"key": {
		// 		"key": true
		// 	}
		// }
	}`;

	expect(uncomment(input)).toMatchSnapshot();
});

test("nested arrays", () => {
	const input = `{
		// @foo bar=5
		// "key": [
		// 	[
		// 		{ "key": true }
		// 	]
		// ]
	}`;

	expect(uncomment(input)).toMatchSnapshot();
});

test("object in array", () => {
	const input = `{
		"key": [
			// @foo bar=5
			// {
			// 	"key": true
			// },
			{
				"key": true
			}
		]
	}`;

	expect(uncomment(input)).toMatchSnapshot();
});

test("partially uncommented", () => {
	const input = `{
		// @foo bar=5
		// "key": true,

		// @foo bar=6
		"key2": false
	}`;

	expect(uncomment(input)).toMatchSnapshot();
});

test("4-space indentation", () => {
	const input = `{
			// @foo bar=5
			// "key": true,

			// @foo bar=6
			// "key2": [
			// 	{ "key": true }
			// ]
	}`;

	expect(uncomment(input)).toMatchSnapshot();
});

test("mixed indentation", () => {
	const input = `{
		// @foo bar=5
		// "key": true,

			// @foo bar=6
			// "key2": [
			//   { "key": true }
			// ]
	}`;

	expect(uncomment(input)).toMatchSnapshot();
});

test("no indentation", () => {
	const input = `{
// @foo bar=5
// "key": true,

// @foo bar=6
// "key2": [
// { "key": true }
// ]
}`;

	expect(uncomment(input)).toMatchSnapshot();
});
