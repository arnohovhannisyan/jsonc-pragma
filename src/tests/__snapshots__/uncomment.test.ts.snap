// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`4-space indentation 1`] = `
"{
			// @foo bar=5
			\\"key\\": true,

			// @foo bar=6
			\\"key2\\": [
				{ \\"key\\": true }
			]
	}"
`;

exports[`arrays 1`] = `
"{
		// @foo bar=5
		\\"key\\": [
			{ \\"key\\": true }
		]
	}"
`;

exports[`basic functionality 1`] = `
"{
		// @foo bar=5
		\\"key\\": true
	}"
`;

exports[`mixed indentation 1`] = `
"{
		// @foo bar=5
		\\"key\\": true,

			// @foo bar=6
			\\"key2\\": [
			  { \\"key\\": true }
			]
	}"
`;

exports[`multiple pragmas 1`] = `
"{
		// @foo bar=5
		\\"key\\": true,

		// @baz bar=10
		\\"key\\": true
	}"
`;

exports[`nested arrays 1`] = `
"{
		// @foo bar=5
		\\"key\\": [
			[
				{ \\"key\\": true }
			]
		]
	}"
`;

exports[`nested objects 1`] = `
"{
		// @foo bar=5
		\\"key\\": {
			\\"key\\": {
				\\"key\\": true
			}
		}
	}"
`;

exports[`no indentation 1`] = `
"{
// @foo bar=5
\\"key\\": true,

// @foo bar=6
\\"key2\\": [
{ \\"key\\": true }
]
}"
`;

exports[`object in array 1`] = `
"{
		\\"key\\": [
			// @foo bar=5
			{
				\\"key\\": true
			},
			{
				\\"key\\": true
			}
		]
	}"
`;

exports[`objects 1`] = `
"{
		// @foo bar=5
		\\"key\\": {
			\\"key\\": true
		}
	}"
`;

exports[`partially uncommented 1`] = `
"{
		// @foo bar=5
		\\"key\\": true,

		// @foo bar=6
		\\"key2\\": false
	}"
`;

exports[`selectors 1`] = `
"{
		// @foo bar=5
		// \\"key\\": true,

		// @baz bar=10
		\\"key\\": true
	}"
`;
